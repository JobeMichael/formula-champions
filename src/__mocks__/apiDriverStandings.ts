export const apiDriverStandings = {
  MRData: {
    StandingsTable: {
      season: "2006",
      StandingsLists: [
        {
          season: "2006",
          DriverStandings: [
            {
              position: "1",
              points: "134",
              wins: "7",
              Driver: {
                driverId: "alonso",
                code: "ALO",
                givenName: "Fernando",
                familyName: "Alonso",
                nationality: "Spanish",
              },
              Constructors: [
                {
                  name: "Renault",
                },
              ],
            },
          ],
        },
      ],
    },
  },
};
