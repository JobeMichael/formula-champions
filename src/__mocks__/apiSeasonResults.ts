export const apiSeasonResults = {
  MRData: {
    xmlns: "http://ergast.com/mrd/1.4",
    series: "f1",
    url: "http://ergast.com/api/f1/2005/results/1.json",
    limit: "30",
    offset: "0",
    total: "19",
    RaceTable: {
      season: "2005",
      position: "1",
      Races: [
        {
          season: "2005",
          round: "1",
          url: "http://en.wikipedia.org/wiki/2005_Australian_Grand_Prix",
          raceName: "Australian Grand Prix",
          Circuit: {
            circuitId: "albert_park",
            url: "http://en.wikipedia.org/wiki/Melbourne_Grand_Prix_Circuit",
            circuitName: "Albert Park Grand Prix Circuit",
            Location: {
              lat: "-37.8497",
              long: "144.968",
              locality: "Melbourne",
              country: "Australia",
            },
          },
          date: "2005-03-06",
          time: "14:00:00Z",
          Results: [
            {
              number: "6",
              position: "1",
              positionText: "1",
              points: "10",
              Driver: {
                driverId: "fisichella",
                code: "FIS",
                url: "http://en.wikipedia.org/wiki/Giancarlo_Fisichella",
                givenName: "Giancarlo",
                familyName: "Fisichella",
                dateOfBirth: "1973-01-14",
                nationality: "Italian",
              },
              Constructor: {
                constructorId: "renault",
                url: "http://en.wikipedia.org/wiki/Renault_in_Formula_One",
                name: "Renault",
                nationality: "French",
              },
              grid: "1",
              laps: "57",
              status: "Finished",
              Time: {
                millis: "5057336",
                time: "1:24:17.336",
              },
              FastestLap: {
                rank: "2",
                lap: "55",
                Time: {
                  time: "1:25.994",
                },
              },
            },
          ],
        },
      ],
    },
  },
};
