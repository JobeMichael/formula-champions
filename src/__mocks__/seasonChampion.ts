export const seasonChampion = {
  season: "2006",
  points: "134",
  wins: "7",
  name: "Fernando Alonso",
  nationality: "Spanish",
  team: "Renault",
  driverId: "alonso",
};
