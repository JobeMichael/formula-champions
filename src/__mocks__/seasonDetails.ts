export const seasonDetails = {
  date: "2005-03-06",
  name: "Giancarlo Fisichella",
  nationality: "Italian",
  raceName: "Australian Grand Prix",
  season: "2005",
  team: "Renault",
  time: "1:24:17.336",
  driverId: "fisichella",
};
