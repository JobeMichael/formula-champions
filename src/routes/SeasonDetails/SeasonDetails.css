.seasons-wrapper {
  width: 100%;
  padding: 0 30px;
}
.seasons-header {
  border: 1px solid #ccc;
  margin-bottom: 20px;
  display: flex;
  flex-direction: column;
  padding: 10px 0;
  justify-content: center;
  align-items: center;
}
.seasons-wrapper .table-responsive {
  margin-top: 15px;
}

.seasons-wrapper a button {
  background-color: #2c8280;
  border: none;
  font-weight: 600;
}
