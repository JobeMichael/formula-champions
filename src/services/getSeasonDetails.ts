import httpClient from "./httpClient";

interface GetSeasonDetails {
  (year: number): Promise<Array<SeasonDetails>>;
}

export interface SeasonDetails {
  name: string;
  nationality: string;
  points: string;
  season: string;
  team: string;
  time: string;
  raceName: string;
  driverId: number;
}

const getSeasonDetails = async (year: string) => {
  const { instance } = httpClient();
  const url = `${year}/results/1.json`;
  const response = await instance
    .get(url)
    .catch((error) => console.log("Error!"));

  if (response) {
    return getFormattedData(response.data.MRData.RaceTable.Races);
  }
  return [];
};

const getFormattedData = <T>(data: T[]): Array<T> =>
  data.reduce((acc: any = [], item: any) => {
    const {
      season,
      date,
      raceName,
      Results: [
        {
          Driver: { familyName, givenName, nationality, driverId },
          Constructor: { name: team },
          Time: { time },
        },
      ],
    } = item;

    acc.push({
      name: `${givenName} ${familyName}`,
      season,
      nationality,
      date,
      raceName,
      team,
      time,
      driverId,
    });

    return acc;
  }, []);

export default getSeasonDetails;
