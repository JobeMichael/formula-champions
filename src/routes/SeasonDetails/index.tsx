import React, { useEffect, useState } from "react";
import { Button } from "react-bootstrap";
import { useParams } from "react-router-dom";
import SeasonRaces from "../../components/SeasonRaces";
import Link from "../../components/UI/Link";
import getSeasonDetails from "../../services/getSeasonDetails";

const Details = () => {
  const [seasonDetails, setSeasonDetails] = useState() as any;
  console.log("Details -> seasonDetails", seasonDetails);
  const { year } = useParams();

  useEffect(() => {
    const fetchSeasonsData = async () => {
      if (year) {
        const result = await getSeasonDetails(year);
        setSeasonDetails(result);
      }
    };
    fetchSeasonsData();
  }, [setSeasonDetails, year]);

  return (
    <div>
      <Link to="/">
        <Button variant="link">Back to list</Button>
      </Link>
      {seasonDetails && <SeasonRaces data={seasonDetails} />}
    </div>
  );
};

export default Details;
