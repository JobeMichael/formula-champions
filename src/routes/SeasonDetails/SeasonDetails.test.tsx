import { render, waitForElement } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import SeasonDetails from ".";
import getSeasonDetails from "../../services/getSeasonDetails";
import { seasonDetails } from "../../__mocks__/seasonDetails";

jest.mock("../../services/getSeasonDetails.ts");

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"), // use actual for all non-hook parts
  useParams: () => ({
    year: "2005",
  }),
}));

describe("<SeasonDetails/>", () => {
  it("renders SeasonDetails correctly", async () => {
    getSeasonDetails.mockReturnValue([seasonDetails]);

    const { asFragment, getByText } = render(
      <MemoryRouter>
        <SeasonDetails />
      </MemoryRouter>
    );

    await waitForElement(() => getByText("Fisichella Giancarlo"));
    expect(asFragment()).toMatchSnapshot();
  });
});
