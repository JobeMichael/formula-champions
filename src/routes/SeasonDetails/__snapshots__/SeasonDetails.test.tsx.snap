// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SeasonDetails/> renders SeasonDetails correctly 1`] = `
<DocumentFragment>
  <div
    class="seasons-wrapper"
  >
    <div
      class="seasons-header"
    >
      <h1>
        2005 Season
      </h1>
      <p />
    </div>
    <a
      href="/"
    >
      <button
        class="btn btn-info"
        type="button"
      >
        &lt; Back to list
      </button>
    </a>
    <div
      class="table-responsive"
    >
      <table
        class="table table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              Race
            </th>
            <th>
              Driver
            </th>
            <th>
              Team
            </th>
            <th>
              Nationality
            </th>
            <th>
              Time
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              Australian Grand Prix
            </td>
            <td>
              Giancarlo Fisichella
            </td>
            <td>
              Renault
            </td>
            <td>
              Italian
            </td>
            <td>
              1:24:17.336
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
